# **DATA ANALYSIS - "PRODUCT_DATA"**

#### Importing packages

In [1]:

import random
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from IPython.display import clear_output

#### Importing chardert to read CSV with encoding -8

In [4]:
pip install chardet

Note: you may need to restart the kernel to use updated packages.


In [3]:
import chardet
# Detect the encoding
with open('/Users/aleksandarkling/Downloads/Raw Data/Python/product_data.csv', 'rb') as rawdata:
    result = chardet.detect(rawdata.read(10000))  # Read the first 10,000 bytes
    encoding = result['encoding']

ModuleNotFoundError: No module named 'chardet'

#### Importing product_data.csv file

In [9]:
# Load the file using the detected encoding
# df = pd.read_csv(r'/Users/aleksandarkling/Downloads/Raw Data/Python/product_data.csv',encoding=encoding)
df = pd.read_csv(r'/Users/ANDRES/Documents/GitHub/OBETAWarehousingAnalyticsProject/Data/002 product_data.csv', encoding="latin", header=None)

In [12]:
df.head()

product_id                    product_desc               product_group
0     000052        PUNCH II MSW1 1500mm PUN                 35_Leuchten
1    1036628              H05VV-F2X1 5WS 50M           16_Sonderverkäufe
2    1052053                 H07RN-F3G1 100M           16_Sonderverkäufe
3     110109  SIEM DELTA Doppelta 2S 5TD2111  32_Schalter_Steckvorrichtg
4     110125  SIEM PLUS Wip Univ ews 5TG7581  32_Schalter_Steckvorrichtg

In [13]:
df[df.product_desc.isna()]

product_id product_desc                  product_group
9328        256092          NaN  31_Install.-Befestigungs-Mat.
13246       363612          NaN                37_Leuchtmittel
14712      1849999          NaN             42_Netzwerktechnik
16676       470240          NaN      18_Haustechnik_Hausgeräte
19138       470239          NaN      18_Haustechnik_Hausgeräte
...            ...          ...                            ...
1999494     Z02686          NaN             42_Netzwerktechnik
2001975     Z02687          NaN             42_Netzwerktechnik
2014789     Z02685          NaN             42_Netzwerktechnik
2021922     Z02688          NaN             42_Netzwerktechnik
2131610     X26375          NaN       40_Sprechen_Schwachstrom

[6023 rows x 3 columns]

In [15]:
duplicates_pick_date = df.duplicated(subset = list(df.columns))
df[duplicates_pick_date].head()

Empty DataFrame
Columns: [product_id, product_desc, product_group]
Index: []

In [18]:
# check the duplicated cases for a specific row giving a dataframe and the columns to evaluate
def check_duplicates_row_columns(pick_data, index_to_check, columns_to_evaluate):

    index_row_pick_data = pick_data.loc[index_to_check, columns_to_evaluate]

    matching_rows = pick_data[pick_data[columns_to_evaluate].eq(index_row_pick_data).all(axis=1)]
    
    return(matching_rows)

In [23]:
duplicates_pick_date = df.duplicated(subset = ["product_id"])
df[duplicates_pick_date].head()

Empty DataFrame
Columns: [product_id, product_desc, product_group]
Index: []

In [ ]:
duplicates_pick_date = df.duplicated(subset = ["product_desc"])
df[duplicates_pick_date].head()

product_id                    product_desc           product_group
86      000395    .Mondia G2 WD2 LED1600-840 E  36_Entladungs-Leuchten
217     000420   .LED Einbau Downlight HZ Rund             35_Leuchten
374     000034      ArimoS M84 CDP LED4000-840  36_Entladungs-Leuchten
379     000248   .TRILUX Siella 31W/840 3400lm  36_Entladungs-Leuchten
443     000437  .LED Einbau Downlight Slim Eco             35_Leuchten

In [22]:
check_duplicates_row_columns(df, 374, ["product_desc"])

product_id                product_desc           product_group
325         000038  ArimoS M84 CDP LED4000-840  36_Entladungs-Leuchten
374         000034  ArimoS M84 CDP LED4000-840  36_Entladungs-Leuchten
1610883     U32594  ArimoS M84 CDP LED4000-840  36_Entladungs-Leuchten
1630151     U32595  ArimoS M84 CDP LED4000-840  36_Entladungs-Leuchten
1895133     X01030  ArimoS M84 CDP LED4000-840  36_Entladungs-Leuchten
1902450     X01029  ArimoS M84 CDP LED4000-840  36_Entladungs-Leuchten

In [81]:
##product= pd.read_csv("product_name.csv",header=None,encoding='latin-1')

#### Adding column names

In [11]:
df.columns=["product_id","product_desc","product_group"]

#### Using describe function to get top-line numerical data for each column

In [11]:
df.describe()

product_id       product_desc    product_group
count     2199592            2193569          2199592
unique    2199592            1422887               18
top        Z52127  Klemmenmarkierung  33_Schaltgeräte
freq            1              11251           557163

#### d.type function to get data type of each column (object)

In [23]:
df.dtypes

product_id                   object
product_desc                 object
product_group                object
product_group_number         object
product_group_description    object
dtype: object

#### Detecting missing values in each column. There are 6,023 in the "product_desc" column, none in others. 

In [29]:
df.isna().sum()

product_id                      0
product_desc                 6023
product_group                   0
product_group_number            0
product_group_description       0
dtype: int64

#### Drop missing values. 

In [30]:
df=df.dropna(subset=['product_desc'],axis=0)

#### Re-check for null values. None found. 

In [32]:
df.isna().sum()

product_id                   0
product_desc                 0
product_group                0
product_group_number         0
product_group_description    0
dtype: int64

#### Using groupby function to analyze how many values there are in the data per product_group

In [12]:
df.groupby(["product_group"]).count()

product_id  product_desc
product_group                                          
16_Sonderverkäufe                    5772          5770
18_Haustechnik_Hausgeräte           38959         38955
19_Werkzeug                         79219         79210
2000_Alka_Bürobedarf                78177         78177
20_C-Artikel                          279           279
31_Install.-Befestigungs-Mat.      337184        334304
32_Schalter_Steckvorrichtg         105507        105501
33_Schaltgeräte                    557163        555926
34_Verteiler_Schränke              189348        189307
35_Leuchten                        446853        446777
36_Entladungs-Leuchten             133575        133073
37_Leuchtmittel                     87577         87568
38_Leitungen                         3711          3711
39_Metalle                              8             8
40_Sprechen_Schwachstrom            72889         71650
41_Antenne                          23142         23130
42_Netzwerktechnik                  40228         40222
61_Kataloge                             1             1

In [91]:
##df.to_csv("updated product.csv")

#### Separating product_group column into two separate columns. Newly split columns will be named "product_group_number" and "product group description"

In [19]:
df[['product_group_number', 'product_group_description']] = df["product_group"].str.split('_', n=1, expand=True)


In [20]:
df

product_id                    product_desc  \
0          1036628              H05VV-F2X1 5WS 50M   
1          1052053                 H07RN-F3G1 100M   
2           110109  SIEM DELTA Doppelta 2S 5TD2111   
3           110125  SIEM PLUS Wip Univ ews 5TG7581   
4           110128  SIEM PLUS Wip Kont ews 5TG7580   
...            ...                             ...   
2199587     Z53390     ASJ MD300EB NH-Lasttrennlei   
2199588     Z53392     ASJ MD31B NH-Lasttrennleist   
2199589     Z53394     ASJ MD31HB NH-Lasttrennleis   
2199590     Z50507   ABB LLEG124X560MM24 LED LLEG1   
2199591     Z52127   ABB SLADCG250MM700 LED SLA DC   

                      product_group product_group_number  \
0                 16_Sonderverkäufe                   16   
1                 16_Sonderverkäufe                   16   
2        32_Schalter_Steckvorrichtg                   32   
3        32_Schalter_Steckvorrichtg                   32   
4        32_Schalter_Steckvorrichtg                   32   
...                             ...                  ...   
2199587       34_Verteiler_Schränke                   34   
2199588       34_Verteiler_Schränke                   34   
2199589       34_Verteiler_Schränke                   34   
2199590             37_Leuchtmittel                   37   
2199591             37_Leuchtmittel                   37   

        product_group_description  
0                  Sonderverkäufe  
1                  Sonderverkäufe  
2         Schalter_Steckvorrichtg  
3         Schalter_Steckvorrichtg  
4         Schalter_Steckvorrichtg  
...                           ...  
2199587        Verteiler_Schränke  
2199588        Verteiler_Schränke  
2199589        Verteiler_Schränke  
2199590              Leuchtmittel  
2199591              Leuchtmittel  

[2199592 rows x 5 columns]

In [35]:
df.describe()


product_id       product_desc    product_group product_group_number  \
count     2193569            2193569          2193569              2193569   
unique    2193569            1422887               18                   18   
top        Z52127  Klemmenmarkierung  33_Schaltgeräte                   33   
freq            1              11251           555926               555926   

       product_group_description  
count                    2193569  
unique                        18  
top                 Schaltgeräte  
freq                      555926

In [36]:
print("done :)")

done :)
